### ISSUE TITLE
Include a pow function

### ISSUE THREAD

**Issue by coretemp** [NixOS/nixpkgs#41251]
## Issue description
I have in `math.nix`:

```
let
pow = n : i :
          if i == 1 then n
          else if i == 0 then 1
          else n * pow n (i - 1);
in
{pow = pow; }
```

, but I think this should be part of the standard library under `lib`.

**Comment by volth** [comment]
Recent attempt to add binary shifts (```pow 2``` actually) got stuck in [the discussion](https://github.com/NixOS/nix/pull/2157/files/8a6a14e1f5006543f63117fb1f0a81a0b1024ebd#r187790654) of should the integers remain 64-bits or become arbitrary-large

**Comment by coretemp** [comment]
We could also just call it `pow64`, if that works on 32 bits platforms. 

My concern with this thread is that it would become a bikeshed issue. Even if the name `pow_pow_pow_pow_pow` would be used, that would be better than everyone having their own version of basic functions.

I don't know of the performance overhead of having almost everything in `<math.h>` (or in some existing functional programming language's math library) as primitives, but I would perhaps just generate a corresponding Nix standard library out of that. For documentation one can then just write a link to some other system and there is no extra investment if you already happen to know that other language. To a certain extent this has been done with the list functions already.   

If backwards compatibility is a concern, we could just introduce `lib1.math` and `lib2.math`, and keep the various lib versions around either forever or for a decade.

**Comment by veprbl** [comment]
Speaking of bikeshedding:
```nix
let
  pow = base: exponent:
    if exponent > 0 then
      let
        and1 = x: (x / 2) * 2 != x;
        x = pow base (exponent / 2);
      in
assert and1 0 == false;
assert and1 1 == true;
assert and1 2 == false;
assert and1 3 == true;
        x * x * (if and1 exponent then base else 1)
    else if exponent == 0 then
      1
    else
      throw "undefined";
in
assert pow 0 1000 == 0;
assert pow 1000 0 == 1;
assert pow 2 30 == 1073741824;
assert pow 3 3 == 27;
assert pow (-5) 3 == -125;
  {
    inherit pow;
  }
```

**Comment by volth** [comment]
>  and1 = x: (x / 2) * 2 != x;

```builtins.bitAnd 1``` will be in Nix soon (already in ```master```, not in ```2.0.4```)

**Comment by veprbl** [comment]
@volth Yes. But while we are in transition, nixpkgs would need to implement backup versions in pure nix before we can use those.

**Comment by volth** [comment]
Also both versions may need ```toInt``` or ```assert isInt``` to catch passing float arguments

**Comment by aanderse** [comment]
@coretemp did you end up getting resolution on this, or is it still sitting?

**Comment by monadix** [comment]
"still important to me"

**Comment by dbeecham** [comment]
I would also like this
